daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -613,4 +604,181 @@ object Transaction {
+                  choiceObservers1,
+                  controllersDifferFromActors1,
+                  children1 @ _,
+                  exerciseResult1 @ _,
+                  key1,
+                  byKey1 @ _,
+                ),
+                exe2 @ Node.NodeExercises(
+                  targetCoid2,
+                  templateId2,
+                  choiceId2,
+                  optLocation2 @ _,
+                  consuming2,
+                  actingParties2,
+                  chosenValue2,
+                  stakeholders2,
+                  signatories2,
+                  choiceObservers2,
+                  controllersDifferFromActors2,
+                  children2 @ _,
+                  exerciseResult2 @ _,
+                  key2,
+                  byKey2 @ _,
+                ))
+                // results are checked after the children
+                if targetCoid1 === targetCoid2 &&
+                  templateId1 == templateId2 &&
+                  choiceId1 == choiceId2 &&
+                  consuming1 == consuming2 &&
+                  actingParties1 == actingParties2 &&
+                  chosenValue1 === chosenValue2 &&
+                  stakeholders1 == stakeholders2 &&
+                  signatories1 == signatories2 &&
+                  choiceObservers1 == choiceObservers2 &&
+                  controllersDifferFromActors1 == controllersDifferFromActors2 &&
+                  (key1.isEmpty || key1 === key2) =>
+              loop(
+                children1.iterator.toStream,
+                children2.iterator.toStream,
+                (nid1, exe1, rest1, nid2, exe2, rest2) :: stack
+              )
+            case (
+                Node.NodeLookupByKey(templateId1, optLocation1 @ _, key1, result1),
+                Node.NodeLookupByKey(templateId2, optLocation2 @ _, key2, result2)
+                )
+                if templateId1 == templateId2 &&
+                  key1 === key2 &&
+                  result1 === result2 =>
+              loop(rest1, rest2, stack)
+            case _ =>
+              fail(Some(nid1), Some(nid2))
+          }
+
+        case (Stream.Empty, Stream.Empty) =>
+          stack match {
+            case (nid1, exe1, nids1, nid2, exe2, nids2) :: rest =>
+              if (exe1.exerciseResult.isEmpty || exe1.exerciseResult === exe2.exerciseResult)
+                loop(nids1, nids2, rest)
+              else
+                fail(Some(nid1), Some(nid2))
+            case Nil =>
+              Right(())
+          }
+
+        case _ =>
+          fail(nids1.headOption, nids2.headOption)
+
+      }
+
+    loop(recorded.roots.iterator.toStream, replayed.roots.iterator.toStream)
+
+  }
+
+}
+
+case class ReplayMismatch[Nid, Cid, Val](
+    recorded: GenTransaction[Nid, Cid, Val],
+    recordedMismatch: Option[Nid],
+    replayed: GenTransaction[Nid, Cid, Val],
+    replayedMismatch: Option[Nid]
+) {
+  require(recordedMismatch.isDefined || replayedMismatch.isDefined)
```

Comment:
Is there no way of just accepting `Nid`s instead of `Option[Nid]`s?

Reply:
Using `Nid` instead of `Option[Nid]` will oblige us to defined 3 different Mismatch Error case classes: 
- One for the case where a recorded node is different from a replayed node
- One for the case a recorded node does not map to any replayed node (e.g. consider comparing a transaction whit 3 root nodes  with a transaction containing only 2 root nodes)
- One for the case a replayed node does not map any record node (symmetric of the previous case).

Do you think it is worthwhile ? 

@andreaslochbihler-da, as the primary consumer of this case class can you tell me if you prefer to have the 3 different case classes.

Reply:
I imagine some usage like the following:

```
case ReplayMismatch(Some(recorded) Some(replayed)) =>
case ReplayMismatch(None, Some(replayed)) =>
case ReplayMismatch(Some(recorded), None) =>
```

vs. 

```
case ReplayNodeMismatch(recorded, replayed) =>
case RecordedNodeMissing(replayed) =>
case ReplayedNodeMissing(recorded) =>
```

The version with the three case classes is IMO a bit more self-explanatory, but I don't have a strong opinion.

Reply:
I guess the version with the three case classes is better because you don't need to even think about the `(None, None)` case.

Reply:
I integrate the version with 3 case classes.